File: app/src/main/java/com/arn/ytscrobble/SessListener.java
package com.arn.ytscrobble;

import android.content.Context;
import android.content.SharedPreferences;
import android.media.MediaMetadata;
import android.media.session.MediaController;
import android.media.session.MediaController.Callback;
import android.media.session.MediaSession;
import android.media.session.MediaSessionManager.OnActiveSessionsChangedListener;
import android.media.session.PlaybackState;
import android.support.annotation.NonNull;
import android.preference.PreferenceManager;
import android.util.Pair;

import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Created by arn on 04/07/2017.
 */

class SessListener implements OnActiveSessionsChangedListener{
    private SharedPreferences pref=null;
    private Context c;
    private NLService.ScrobbleHandler handler;
    private YtCallback ytCallback = new YtCallback();

    private final Map<MediaSession.Token, Pair<MediaController, Callback>> mControllers = new HashMap<>();


    SessListener(Context c, NLService.ScrobbleHandler h){
        this.c = c;
        handler = h;
        pref = PreferenceManager.getDefaultSharedPreferences(c);
    }

    @Override
    public void onActiveSessionsChanged(List<MediaController> controllers) {
        int controllerCount = ((null == controllers) ? 0 : controllers.size());
        Set<MediaSession.Token> tokens = new HashSet<>(controllerCount);
        for (int i = 0; i < controllerCount; i++) {
            MediaController controller = controllers.get(i);
            if (pref.getBoolean("scrobble_youtube", true) &&
                    controller.getPackageName().equals( NLService.YOUTUBE_PACKAGE)) {
                tokens.add(controller.getSessionToken());
                // Only add tokens that we don't already have.
                if (!mControllers.containsKey(controller.getSessionToken())) {
                    controller.registerCallback(ytCallback);
                    Pair<MediaController, Callback> pair = Pair.create(controller, (Callback)ytCallback);
                    synchronized (mControllers) {
                        mControllers.put(controller.getSessionToken(), pair);
                    }
                }
            }
        }
        // Now remove old sessions that are not longer active.
        for (Map.Entry<MediaSession.Token, Pair<MediaController, Callback>> entry : mControllers.entrySet()) {
            MediaSession.Token token = entry.getKey();
            if (!tokens.contains(token)) {
                Pair<MediaController, MediaController.Callback> pair = entry.getValue();
                pair.first.unregisterCallback(pair.second);
                synchronized (mControllers) {
                    mControllers.remove(token);
                    handler.remove(ytCallback.lastHash);
                }
            }
        }
    }

    private class YtCallback extends Callback {
        MediaMetadata metadata= null;
        int lastHash = 0;
        long lastPos = 1;

        @Override
        public void onMetadataChanged(MediaMetadata metadata) {
            super.onMetadataChanged(metadata);
            Stuff.log(c, "metadata changed " );
            this.metadata = metadata;
            lastPos = 1;
        }

        @Override
        public void onPlaybackStateChanged(@NonNull PlaybackState state) {
            super.onPlaybackStateChanged(state);
            if (metadata == null)
                return;
            String title = metadata.getString(MediaMetadata.METADATA_KEY_TITLE);
//                String artist =  metadata.getString(MediaMetadata.METADATA_KEY_ARTIST);

            if (title.equals(""))
                return;
            if (state.getState()==PlaybackState.STATE_PAUSED) {
//                    cancel scrobbling if within time
                lastPos = state.getPosition();
                handler.remove(lastHash);
                Stuff.log(c, "paused");
            } else if ( state.getState()==PlaybackState.STATE_STOPPED){
                // a replay should count as another scrobble
                lastPos = 1;
                handler.remove(lastHash);
                Stuff.log(c, "stopped");
            } else if (state.getState()==PlaybackState.STATE_PLAYING || state.getState()==PlaybackState.STATE_BUFFERING){
                if (state.getState()==PlaybackState.STATE_BUFFERING && state.getPosition() == 0)
                    return; //dont scrobble first buffering

                Stuff.log(c, "playing: "+ state.getPosition()+ " < " + lastPos + " " + title);
                if (pref.getBoolean("scrobble_youtube", true) && state.getPosition() < lastPos || lastPos == 1) {
//                    lastPos = state.getPosition();
                    lastHash = handler.scrobble(title);
                }
            } else if (state.getState()==PlaybackState.STATE_CONNECTING){
                Stuff.log(c, "connecting "+ state.getPosition());
            } else
                Stuff.log(c, "other ("+state.getState()+") : " + title);

        }
    }
}

File: composeApp/src/androidMain/kotlin/com/arn/scrobble/media/SessListener.kt
package com.arn.scrobble.media

import android.annotation.SuppressLint
import android.content.ComponentName
import android.media.AudioManager
import android.media.MediaMetadata
import android.media.session.MediaController
import android.media.session.MediaSessionManager.OnActiveSessionsChangedListener
import android.media.session.PlaybackState
import android.os.Build
import android.os.Bundle
import android.service.quicksettings.TileService
import co.touchlab.kermit.Logger
import com.arn.scrobble.BuildKonfig
import com.arn.scrobble.MasterSwitchQS
import com.arn.scrobble.media.PlayerActions.love
import com.arn.scrobble.media.PlayerActions.unlove
import com.arn.scrobble.utils.AndroidStuff
import com.arn.scrobble.utils.AndroidStuff.dump
import com.arn.scrobble.utils.AndroidStuff.toast
import com.arn.scrobble.utils.PlatformStuff
import com.arn.scrobble.utils.Stuff.stateInWithCache
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.launch
import org.jetbrains.compose.resources.getString
import pano_scrobbler.composeapp.generated.resources.Res
import pano_scrobbler.composeapp.generated.resources.mute
import pano_scrobbler.composeapp.generated.resources.skip

class SessListener(
    scope: CoroutineScope,
    scrobbleQueue: ScrobbleQueue,
    private val audioManager: AudioManager,
) : MediaListener(scope, scrobbleQueue), OnActiveSessionsChangedListener {

    override val notifyTimelineUpdates = false
    private val mainPrefs = PlatformStuff.mainPrefs
    private var platformControllers = emptyList<MediaController>()

    private val scrobbleSpotifyRemote =
        mainPrefs.data.stateInWithCache(scope) { it.scrobbleSpotifyRemoteP }

    private val autoDetectApps =
        mainPrefs.data.stateInWithCache(scope) { it.autoDetectAppsP }

    private val blockedPackages =
        mainPrefs.data.stateInWithCache(scope) { it.blockedPackages }

    init {
        scope.launch {
            combine(
                allowedPackages,
                blockedPackages,
                autoDetectApps,
                scrobblerEnabled,
            ) { allowed, blocked, autoDetect, scrobbleEnabled ->
                onActiveSessionsChanged(platformControllers)
                val tokensToKeep = sessionTrackers
                    .filter { (k, v) ->
                        shouldScrobble(v.trackInfo.appId)
                    }
                    .keys
                removeSessions(tokensToKeep)
            }.collect()
        }

        scope.launch {
            scrobblerEnabled.collectLatest {
                try {
                    TileService.requestListeningState(
                        AndroidStuff.applicationContext,
                        ComponentName(AndroidStuff.applicationContext, MasterSwitchQS::class.java)
                    )
                } catch (e: Exception) {
                    Logger.e(e) { "Failed to update QS tile state" }
                }
            }
        }
    }

    override fun shouldScrobble(rawAppId: String): Boolean {
        val should = scrobblerEnabled.value &&
                (rawAppId in allowedPackages.value ||
                        (autoDetectApps.value && rawAppId !in blockedPackages.value))
        return should
    }

    fun isAppAllowListed(appId: String): Boolean {
        return appId in allowedPackages.value
    }

    @Synchronized
    override fun onActiveSessionsChanged(controllers: List<MediaController>?) {

        Logger.i {
            "controllers: " + controllers?.joinToString {
                "${it.packageName}|${it.tagCompat}@" + it.sessionToken.hashCode().toHexString()
            }
        }

        if (controllers == null)
            return

        this.platformControllers = controllers

        if (!scrobblerEnabled.value)
            return

        val controllersFiltered = controllers.mapNotNull {
            if (shouldScrobble(it.packageName) &&
                it.sessionToken !in sessionTrackers
            )
                MediaController(AndroidStuff.applicationContext, it.sessionToken)
            else null
        }

//        val tokens = mutableSetOf<MediaSession.Token>()
        for (controller in controllersFiltered) {
//            if (shouldScrobble(controller.packageName)) {
//                tokens.add(controller.sessionToken) // Only add tokens that we don't already have.
//                if (controller.sessionToken !in controllersMap) {

            val appId = controller.packageName
            val sessionId = controller.tagCompat

            val uniqueId = "$appId|$sessionId@" + controller.sessionToken.hashCode().toHexString()

            val playingTrackInfo = createTrackInfo(controller.packageName, uniqueId)

            val sessionTracker = AndroidSessionTracker(controller, playingTrackInfo)

            controller.registerCallback(sessionTracker.callback)

            sessionTracker.callback.onPlaybackStateChanged(controller.playbackState)
            sessionTracker.callback.onMetadataChanged(controller.metadata)

            if (BuildKonfig.DEBUG) {
                sessionTracker.callback.onExtrasChanged(controller.extras)
                sessionTracker.callback.onAudioInfoChanged(controller.playbackInfo)
            }

            sessionTrackers[controller.sessionToken] = sessionTracker
        }

        // Now remove old sessions that are no longer active.
//        removeSessions(tokens)
    }

    override fun isMediaPlaying() = platformControllers.any { it.isMediaPlaying() }

    private val MediaController.tagCompat
        @SuppressLint("NewApi")
        get() = when {
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.R -> tag
            Build.VERSION.SDK_INT <= Build.VERSION_CODES.O_MR1 -> tag
            else -> "greylist-max-o"
        }

    override fun mute(hash: Int) {
        // if pano didnt mute this, dont unmute later
        if (mutedHash == null && audioManager.isStreamMute(AudioManager.STREAM_MUSIC))
            return

        val tracker = findTrackerByHash(hash)
        if (tracker != null) {
            audioManager.adjustStreamVolume(
                AudioManager.STREAM_MUSIC,
                AudioManager.ADJUST_MUTE,
                0
            )
            Logger.i { "mute: done" }

            mutedHash = hash
            tracker.isMuted = true

            scope.launch(Dispatchers.Main) {
                AndroidStuff.applicationContext.toast(getString(Res.string.mute))
            }
        }
    }

    override fun unmute(clearMutedHash: Boolean) {
        if (mutedHash != null) {
            audioManager.adjustStreamVolume(
                AudioManager.STREAM_MUSIC,
                AudioManager.ADJUST_UNMUTE,
                0
            )
            Logger.i { "unmute: done" }

            val tracker = findTrackerByHash(mutedHash!!)
            tracker?.isMuted = false

            if (clearMutedHash)
                mutedHash = null
        }
    }


    //        Build.VERSION.SDK_INT >= Build.VERSION_CODES.R &&
    // MediaController.getTag() exists on Android 10 and lower but is marked as @hide

    private fun MediaController.isMediaPlaying() =
        playbackState?.state in arrayOf(
            PlaybackState.STATE_PLAYING,
            PlaybackState.STATE_BUFFERING
        ) && !metadata?.getString(MediaMetadata.METADATA_KEY_TITLE).isNullOrEmpty()

    inner class AndroidSessionTracker(
        val controller: MediaController,
        trackInfo: PlayingTrackInfo,
    ) : SessionTracker(trackInfo) {

        override fun skip() {
            controller.transportControls.skipToNext()

            scope.launch(Dispatchers.Main) {
                AndroidStuff.applicationContext.toast(getString(Res.string.skip))
            }
        }

        override fun love() {
            controller.love()
        }

        override fun unlove() {
            controller.unlove()
        }

        override fun stop() {
            pause()
            controller.unregisterCallback(callback)
        }

        val callback = object : MediaController.Callback() {

            init {
                controller.registerCallback(this)
            }

            @Synchronized
            override fun onMetadataChanged(metadata: MediaMetadata?) {
                metadata ?: return


//            if (BuildKonfig.DEBUG)
//                metadata.dump()

                val (metadataInfo, ignoreScrobble) = transformMediaMetadata(trackInfo, metadata)

                metadataChanged(metadataInfo, ignoreScrobble)
            }

            @Synchronized
            override fun onPlaybackStateChanged(state: PlaybackState?) {
                state ?: return

                val options = TransformMetadataOptions(
                    scrobbleSpotifyRemote = scrobbleSpotifyRemote.value
                )
                val (playbackInfo, ignoreScrobble) = transformPlaybackState(
                    trackInfo,
                    state,
                    options
                )

                playbackStateChanged(playbackInfo, ignoreScrobble)
            }

            override fun onSessionDestroyed() {
                Logger.d { "onSessionDestroyed ${trackInfo.appId}" }
                stop()
            }

            override fun onExtrasChanged(extras: Bundle?) {
                Logger.d { "extras updated ${trackInfo.appId}: ${extras.dump()}" }
            }

            override fun onSessionEvent(event: String, extras: Bundle?) {
                Logger.d { "onSessionEvent ${trackInfo.appId}: $event ${extras.dump()}" }
            }

            override fun onAudioInfoChanged(info: MediaController.PlaybackInfo) {
                Logger.d { "audioinfo updated ${trackInfo.appId}: $info" }
            }
        }
    }
}
